kythe/cxx/indexer/textproto/recordio_textparser.cc

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2023 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <sstream>
+
+#include "absl/strings/ascii.h"
+#include "absl/strings/match.h"
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+#include "glog/logging.h"
+#include "google/protobuf/descriptor.h"
+#include "google/protobuf/dynamic_message.h"
+#include "google/protobuf/text_format.h"
+
+namespace kythe {
+namespace lang_textproto {
+
+void ParseRecordioChunks(
+    std::string content, absl::optional<std::string> record_delimiter,
+    google::protobuf::DynamicMessageFactory& factory,
+    const google::protobuf::Descriptor* descriptor,
+    std::function<void(google::protobuf::Message*,
+                       google::protobuf::TextFormat::ParseInfoTree&, int)>
+        callback) {
+  std::string delimiter = "";
+  if (record_delimiter.has_value()) {
+    delimiter = absl::StripAsciiWhitespace(record_delimiter.value());
+  }
+  // TODO(chuthan): Accept parser as input instead of creating one here.
+  google::protobuf::TextFormat::Parser parser;
+  // Parse textproto into @proto, recording input locations to @parse_tree.
+  google::protobuf::TextFormat::ParseInfoTree parse_tree;
+  parser.WriteLocationsTo(&parse_tree);
+  // Relax parser restrictions - even if the proto is partially
+  // ill-defined, we'd like to analyze the parts that are good.
+  parser.AllowPartialMessage(true);
+  parser.AllowUnknownExtension(true);
+
+  std::istringstream stream(content);
+  std::ostringstream chunk;
+  std::string line;
+  int num_of_lines = 0;
+  int chunk_begin_line = 0;
+  bool textprto_chunk_read = false;
+  while (std::getline(stream, line)) {
```

Comment:
This is largely just `absl::StrSplit(content, absl::StrCat("\n", delimiter, "\n"))` with unnecessary copies, but I agree with Jay that this should be pulled into the main loop.

Reply:
When the separator is new_line, I still need to handle the case where comments are separated by new-line. So I need to check if each chunk only contains comments/whitespaces and then make it part of next chunk.

Also, one of the recordio-textformatted files within google3 has 70K and next few are around 30K lines. 
Created some references to files within this internal document go/github-kythe-pr5529

What do you mean this should be pulled into the main loop?

Reply:
By "Pulled into the main loop" I mean that the multi-chunk and extant "single-chunk" logic should not be duplicated as they are here.

Yes, this is why I say "largely". As written, the use of `std::getline` could be replaced by `absl::StrSplit(contents, ByChar('\n'))` for "free".

Reply:
Thanks, I've tried to make changes to reflect what you suggested. I've made the function split into chunks and return that.. rather than also trying to convert it to proto object.